plugins/org.eclipse.n4js.ide/src/org/eclipse/n4js/ide/editor/contentassist/NodeModelTokenSource.java

```diff
@@ -50,20 +59,112 @@
+	 * to reduce the number of tokens.
+	 */
 
-	NodeModelTokenSource(INode node, ContentAssistTokenTypeMapper tokenTypeMapper, N4JSGrammarAccess grammarAccess) {
-		this(node, 0, Integer.MAX_VALUE, tokenTypeMapper, grammarAccess);
+	NodeModelTokenSource(INode node, ContentAssistTokenTypeMapper tokenTypeMapper, N4JSGrammarAccess grammarAccess,
+			boolean filter) {
+		this(node, 0, Integer.MAX_VALUE, tokenTypeMapper, grammarAccess, filter);
 	}
 
 	NodeModelTokenSource(INode node, int startOffset, int endOffset, ContentAssistTokenTypeMapper tokenTypeMapper,
-			N4JSGrammarAccess grammarAccess) {
+			N4JSGrammarAccess grammarAccess, boolean filter) {
 		this.startOffset = startOffset;
 		this.endOffset = endOffset;
 		this.tokenTypeMapper = tokenTypeMapper;
-		this.leafNodes = node.getLeafNodes().iterator();
+
+		this.leafNodes = getLeafIterator(((AbstractNode) node).basicIterator(), filter);
 		this.rightCurlyInBlock = grammarAccess.getBlockAccess().getRightCurlyBracketKeyword_2();
 		this.rightCurlyInArrowExpression = grammarAccess.getArrowExpressionAccess().getRightCurlyBracketKeyword_1_0_2();
 		this.semicolon = grammarAccess.getSemiAccess().getSemicolonKeyword();
+		this.scriptElementCall = grammarAccess.getScriptAccess().getScriptElementsScriptElementParserRuleCall_1_1_0();
+		this.memberDeclarationCall = grammarAccess.getMembersAccess()
+				.getOwnedMembersRawN4MemberDeclarationParserRuleCall_1_0();
+		this.statementsCall = grammarAccess.getBlockAccess().getStatementsStatementParserRuleCall_1_0();
+		this.propertyAssignmentCall1 = grammarAccess.getObjectLiteralAccess()
+				.getPropertyAssignmentsPropertyAssignmentParserRuleCall_2_0_0();
+		this.propertyAssignmentCall2 = grammarAccess.getObjectLiteralAccess()
+				.getPropertyAssignmentsPropertyAssignmentParserRuleCall_2_1_1_0();
+	}
+
+	private Iterator<ILeafNode> getLeafIterator(TreeIterator<AbstractNode> iterator, boolean filter) {
+		return Iterators.filter(filter ? filterIterator(iterator) : iterator, ILeafNode.class);
+	}
+
+	private Iterator<INode> filterIterator(TreeIterator<AbstractNode> iterator) {
+		return new AbstractIterator<>() {
+			@Override
+			protected INode computeNext() {
+				if (iterator.hasNext()) {
+					INode result = iterator.next();
+					if (result instanceof ICompositeNode) {
+						ICompositeNode casted = (ICompositeNode) result;
+						if (casted.getTotalEndOffset() < endOffset - 1) {
+							if (casted.hasChildren() && casted.getLookAhead() == 1) {
+								AbstractElement grammarElement = (AbstractElement) casted.getGrammarElement();
+								if (grammarElement == scriptElementCall || grammarElement == memberDeclarationCall) {
+									INode sibling = casted.getNextSibling();
+									while (sibling instanceof ILeafNode) {
+										ILeafNode siblingLeaf = (ILeafNode) sibling;
+										if (siblingLeaf.isHidden()) {
+											if (siblingLeaf.getTotalEndOffset() >= endOffset) {
+												return result;
```

Comment:
Did you copy that code (meaning body of `filterIterator`) from somewhere?

Reply:
No, this was tailored to the syntax of N4JS. I'd like to generalize this one day to avoid a dependency on grammar elements, but so far it was easier to do it like this.

Reply:
I'll try to add some comments, though.